### PR TITLE
feat(nuxt): `app:rendered` and `app:response` hooks

### LINKED ISSUE #5534: Allow hooking into render context
Sometimes it is necessary (especially for module authors) to hook into the renderer and do modifications or access the rendered HTML **outside of the app lifecycle** (making it possible to extend for SPA mode as well and write code in server context). We can allow this using nitro plugins hooking into [renderer](https://github.com/nuxt/framework/blob/main/packages/nuxt/src/core/runtime/nitro/renderer.ts). I'm proposing two new server/nitro hooks:

- `nuxt:render:rendered (context: NuxtRenderContext)` (Allow the modification and extension to the render result before compiling the template. `html` is unset)
- `nuxt:render:html (html: string, context: NuxtRenderContext)` (read-only access to final rendered html. modification is prohibited by design to avoid heavy string concatenation and preferring small context appending)

Notes:
- Would worth to add `ssrContext` to `RenderResult`.
- We might rename `RenderResult` interface to `NuxtRenderContext` before exposing
- `RenderResult.html` type was never used. I'm proposing to remove it to make types less confusing.
- We might introduce other types of render results (such as `payload`) in the future.

### PR DESCRIPTION
**Pull request by pi0**
### 🔗 Linked issue

Resolves #5534 (compatible with #5565)

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR adds new `nuxt:app:rendered` and `nuxt:app:response` nitro hooks and also provides same render context to the existing `app:rendered` hook usable in Nuxt plugins.

A small additional charge, `X-Powered-By: Nuxt` header is now added to the default response. It can be disabled with nitro plugin and `nuxt:app:response` hook.

Introducing external types for nitro in subsequent PR ~> #6170

**Example usage:** (nitro server plugin)

```js
export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('nuxt:app:rendered', ({ ssrContext, html }) => { console.log({ html }) })
  nitroApp.hooks.hook('nuxt:app:response', ({ response }) => { console.log(response) })
})
```

**Example usage:** (nuxt plugin)

```js
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook('app:rendered', ({ ssrContext, html }) => {
    console.log({ html })
  })
})
```

Example output:

```js
{
  html: {
    htmlAttrs: [ 'data-head-attrs=""' ],
    head: [
      '<title></title><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><meta name="head:count" content="2">',
      '<link rel="modulepreload" href="/_nuxt/@vite/client" as="script" crossorigin><link rel="preload" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-404.vue?vue&type=style&index=0&scoped=8dd555df&lang.css" as="style"><link rel="preload" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-dev.vue?vue&type=style&index=0&scoped=f1f2d31d&lang.css" as="style"><link rel="preload" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-500.vue?vue&type=style&index=0&scoped=979ee0c7&lang.css" as="style"><link rel="modulepreload" href="/_nuxt/entry.mjs" as="script" crossorigin>',
      '<link rel="stylesheet" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-404.vue?vue&type=style&index=0&scoped=8dd555df&lang.css"><link rel="stylesheet" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-dev.vue?vue&type=style&index=0&scoped=f1f2d31d&lang.css"><link rel="stylesheet" href="/_nuxt/@fs/home/pooya/Code/framework/node_modules/@nuxt/ui-templates/dist/templates/error-500.vue?vue&type=style&index=0&scoped=979ee0c7&lang.css">'
    ],
    bodyAttrs: [ 'data-head-attrs=""' ],
    bodyPreprend: [],
    body: [
      '<div id="__nuxt"><!--[--><!-- Edit this file to play around with Nuxt but never commit changes! --><div> Nuxt 3 Playground </div><!--]--></div>'
    ],
    bodyAppend: [
      '<script>window.__NUXT__={data:{},state:{$serror:null},_errors:{},serverRendered:true,config:{public:{},app:{baseURL:"\\u002F",buildAssetsDir:"\\u002F_nuxt\\u002F",cdnURL:""}}}</script>',
      '<script type="module" src="/_nuxt/@vite/client" crossorigin></script><script type="module" src="/_nuxt/entry.mjs" crossorigin></script>'
    ]
  }
}
```

```js
{
  body: '<!DOCTYPE html>...............</html>',
  statusCode: 200,
  statusMessage: undefined,
  headers: { 'Content-Type': 'text/html;charset=UTF-8', 'X-Powered-By': 'Nuxt' }
}
```

### CODE REVIEW

**Review by danielroe — APPROVED** [review]

File: packages/nuxt/src/core/runtime/nitro/renderer.ts [hunk 1]
```diff
@@ -185,3 +193,23 @@ function lazyCachedFunction <T> (fn: () => Promise<T>): () => Promise<T> {
...
+  return meta.filter(Boolean).map(i => i.trim())
+}
+
+function joinTags (meta: string[]) {
+  return meta.join('')
+}
+
+function joinAttrs (meta: string[]) {
+  return meta.join(' ')
+}
```
Comment (danielroe) [inline packages/nuxt/src/core/runtime/nitro/renderer.ts:207]:
```suggestion
function joinTags (tags: string[]) {
  return tags.join('')
}

function joinAttrs (chunks: string[]) {
  return chunks.join(' ')
}
```

File: packages/nuxt/src/core/runtime/nitro/renderer.ts [hunk 2]
```diff
@@ -132,50 +126,64 @@ export default eventHandler(async (event) => {
...
+    html: {
+      htmlAttrs: normalizeMeta([renderedMeta.htmlAttrs]),
+      head: normalizeMeta([
+        renderedMeta.headTags,
+        _rendered.renderResourceHints(),
+        _rendered.renderStyles(),
+        ssrContext.styles
+      ]),
+      bodyAttrs: normalizeMeta([renderedMeta.bodyAttrs]),
+      bodyPreprend: normalizeMeta([
```
Comment (danielroe) [inline packages/nuxt/src/core/runtime/nitro/renderer.ts:156]:
```suggestion
      htmlAttrs: normalizeChunks([renderedMeta.htmlAttrs]),
      head: normalizeChunks([
        renderedMeta.headTags,
        _rendered.renderResourceHints(),
        _rendered.renderStyles(),
        ssrContext.styles
      ]),
      bodyAttrs: normalizeChunks([renderedMeta.bodyAttrs]),
      bodyPreprend: normalizeChunks([
```

File: packages/nuxt/src/core/runtime/nitro/renderer.ts [hunk 3]
```diff
@@ -132,50 +126,64 @@ export default eventHandler(async (event) => {
...
+      bodyAttrs: normalizeChunks([renderedMeta.bodyAttrs]),
+      bodyPreprend: normalizeChunks([
+        renderedMeta.bodyScriptsPrepend,
+        ssrContext.teleports?.body
+      ]),
+      body: [
+      // TODO: Rename to _rendered.body in next vue-bundle-renderer
+        _rendered.html
+      ],
+      bodyAppend: normalizeMeta([
```
Comment (danielroe) [inline packages/nuxt/src/core/runtime/nitro/renderer.ts:164]:
```suggestion
      bodyAppend: normalizeChunks([
```

**Review by danielroe — APPROVED** [review]

**Comment by antfu** [comment]
Love the format! 👍

**Pull request merged**